runtime/gc_glue_java/ScavengerDelegate.hpp

```diff
@@ -0,0 +1,153 @@
+
+	/*
+	 * Scavenger Collector, Private
+	 */
+
+	/**
+	 * Decide if GC percolation should occur for class unloading.
+	 * The class unloading occurs after a user defined number of cycles.
+	 */
+	bool private_shouldPercolateGarbageCollect_classUnloading(MM_EnvironmentBase *envBase);
+
+	/**
+	 * Decide if GC percolation should occur due to active JNI critical
+	 * regions.  Active regions require that objects do not move, which
+	 * prevents scavenging. Percolate collect instead.
+	 */
+	bool private_shouldPercolateGarbageCollect_activeJNICriticalRegions(MM_EnvironmentBase *envBase);
+
+protected:
+public:
+	void masterSetupForGC(MM_EnvironmentBase *env);
+	void workerSetupForGC_clearEnvironmentLangStats(MM_EnvironmentBase *env);
+	void reportScavengeEnd(MM_EnvironmentBase * envBase, bool scavengeSuccessful);
+	void mergeGCStats_mergeLangStats(MM_EnvironmentBase *envBase);
+	void masterThreadGarbageCollect_scavengeComplete(MM_EnvironmentBase *envBase);
+	void masterThreadGarbageCollect_scavengeSuccess(MM_EnvironmentBase *envBase);
+	bool internalGarbageCollect_shouldPercolateGarbageCollect(MM_EnvironmentBase *envBase, PercolateReason *reason, U_32 *gcCode);
+	GC_ObjectScanner *getObjectScanner(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags);
+	void flushReferenceObjects(MM_EnvironmentStandard *env);
+	bool hasIndirectReferentsInNewSpace(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr);
+	bool scavengeIndirectObjectSlots(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr);
+	void backOutIndirectObjectSlots(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr);
+	void backOutIndirectObjects(MM_EnvironmentStandard *env);
+	void reverseForwardedObject(MM_EnvironmentBase *env, MM_ForwardedHeader *forwardedObject);
+#if defined (J9VM_INTERP_COMPRESSED_OBJECT_HEADER)
+	void fixupDestroyedSlot(MM_EnvironmentBase *env, MM_ForwardedHeader *forwardedObject, MM_MemorySubSpaceSemiSpace *subSpaceNew);
+#endif /* J9VM_INTERP_COMPRESSED_OBJECT_HEADER */
+
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	void switchConcurrentForThread(MM_EnvironmentBase *env);
+	void fixupIndirectObjectSlots(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr);
+	bool shouldYield();
+#endif /* OMR_GC_CONCURRENT_SCAVENGER */
+
+	void setShouldScavengeUnfinalizedObjects(bool shouldScavenge) { _shouldScavengeUnfinalizedObjects = shouldScavenge; }
+
+	volatile bool getShouldScavengeFinalizableObjects() { return _shouldScavengeFinalizableObjects; }
+	volatile bool getShouldScavengeUnfinalizedObjects() { return _shouldScavengeUnfinalizedObjects; }
+	volatile bool getShouldScavengeSoftReferenceObjects() { return _shouldScavengeSoftReferenceObjects; }
+	volatile bool getShouldScavengeWeakReferenceObjects() { return _shouldScavengeWeakReferenceObjects; }
+	volatile bool getShouldScavengePhantomReferenceObjects() { return _shouldScavengePhantomReferenceObjects; }
```

Comment:
do these (or any for now) need be virtual?

Reply:
There should be no need to make any functions in the delegate virtual.

Reply:
Unless you meant volatile... The return results were copied from the originals in the CLI. If I understand correctly, using volatile in this way should have no effect, and can be removed. Although, I'd prefer to deliver a change like that in it's own PR, and leave the code as close as possible to the original for this one.

Reply:
yeah, I miss read it. leave it as is